copier/copier.go

```diff
@@ -0,0 +1,1367 @@
+	errorResponse := func(fmtspec string, args ...interface{}) (*Response, func() error, error) {
+		return &Response{Put: PutResponse{Error: fmt.Sprintf(fmtspec, args...)}}, nil, nil
+	}
+	dirUID, dirGID := 0, 0
+	if request.PutOptions.ChownDirs != nil {
+		dirUID, dirGID = request.PutOptions.ChownDirs.UID, request.PutOptions.ChownDirs.GID
+	}
+	dirMode := os.FileMode(0755)
+	if request.PutOptions.ChmodDirs != nil {
+		dirMode = *request.PutOptions.ChmodDirs
+	}
+	var fileUID, fileGID *int
+	if request.PutOptions.ChownFiles != nil {
+		fileUID, fileGID = &request.PutOptions.ChownFiles.UID, &request.PutOptions.ChownFiles.GID
+	}
+	if idMappings != nil && !idMappings.Empty() {
+		containerDirPair := idtools.IDPair{UID: dirUID, GID: dirGID}
+		hostDirPair, err := idMappings.ToHost(containerDirPair)
+		if err != nil {
+			return errorResponse("copier: put: error mapping container filesystem owner %d:%d to host filesystem owners: %v", dirUID, dirGID, err)
+		}
+		dirUID, dirGID = hostDirPair.UID, hostDirPair.GID
+		if request.PutOptions.ChownFiles != nil {
+			containerFilePair := idtools.IDPair{UID: *fileUID, GID: *fileGID}
+			hostFilePair, err := idMappings.ToHost(containerFilePair)
+			if err != nil {
+				return errorResponse("copier: put: error mapping container filesystem owner %d:%d to host filesystem owners: %v", fileUID, fileGID, err)
+			}
+			fileUID, fileGID = &hostFilePair.UID, &hostFilePair.GID
+		}
+	}
+	ensureDirectoryUnderRoot := func(directory string) error {
+		rel, err := filepath.Rel(request.Root, directory)
+		if err != nil {
+			return errors.Wrapf(err, "%q is not a subdirectory of %q", directory, request.Root)
+		}
+		subdir := ""
+		for _, component := range strings.Split(rel, string(os.PathSeparator)) {
+			subdir = filepath.Join(subdir, component)
+			path := filepath.Join(request.Root, subdir)
+			if err := os.Mkdir(path, dirMode); err == nil {
+				err = os.Chown(path, dirUID, dirGID)
+				if err != nil {
+					return errors.Wrapf(err, "copier: put: error setting owner of %q to %d:%d", path, dirUID, dirGID)
+				}
+			} else {
+				if !os.IsExist(err) {
+					return errors.Wrapf(err, "copier: put: error checking directory %q", path)
+				}
+			}
+		}
+		return nil
+	}
+	createFile := func(path string, tr *tar.Reader, mode os.FileMode) (int64, error) {
+		f, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_TRUNC|os.O_EXCL, mode)
+		if err != nil && os.IsExist(err) {
+			if err = os.Remove(path); err != nil {
+				return 0, errors.Wrapf(err, "copier: put: error removing file to be overwritten %q", path)
+			}
+			f, err = os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_TRUNC|os.O_EXCL, mode)
+		}
+		if err != nil {
+			return 0, errors.Wrapf(err, "copier: put: error opening file %q for writing", path)
+		}
+		defer f.Close()
+		n, err := io.Copy(f, tr)
+		if err != nil {
+			return n, errors.Wrapf(err, "copier: put: error writing file %q", path)
+		}
+		if err = f.Chmod(mode); err != nil {
```

Comment:
If this is supposed to be safe against concurrent access, `OpenFile(…, 0600); Chown; Chmod`.

Reply:
Changed it, thanks!